### ISSUE TITLE
Push Notifications are broken.. sometimes.

### ISSUE THREAD

**Issue by lazyadmin111** [ChatSecure/ChatSecure-iOS#637]
Dear Developers,

I installed your nice app at some friends, and I am using it with a own xmpp server (prosody 0.10) + Push-module (cloud_notify). Some of my friends a while back also got some push notifications, so it seemed to work somehow a while ago. But now there aren't any push notifications any more. I wonder now if the push server is down (as my prosody logs seem to indicate, as they say something of "not found"), or if is rather my fault, e.g. because he the push app server doesn't like my self-signed certificate. But if the later, why did it work then sometimes a while ago?

Any helpful comment highly appreciated :-)

greets

**Comment by ge0rg** [comment]
Not a developer, but I can confirm the problem on my public XMPP server (yax.im):

1. Message is sent to the user
2. prosody sends IQ to pubsub.chatsecure.org
3. pubsub.chatsecure.org returns IQ result
4. nothing happens on the smartphone

Debug log from prosody:
```
Jan 03 17:40:34 yax.im:carbons  debug   Sending carbon to user@yax.im/chatsecureXXXX
Jan 03 17:40:34 c2sb659af0      debug   Invoking cloud handle_notify_request for new smacks hibernated stanza...
Jan 03 17:40:34 yax.im:cloud_notify     debug   Sending push notification for user@yax.im to pubsub.chatsecure.org
Jan 03 17:40:34 stanzarouter    debug   Routing to remote...
Jan 03 17:40:34 s2sout3c93a10   debug   going to send stanza to pubsub.chatsecure.org from yax.im
Jan 03 17:40:34 s2sout3c93a10   debug   sending: <iq type='set' to='pubsub.chatsecure.org' from='user@yax.im' id='push'>
Jan 03 17:40:34 s2sout3c93a10   debug   stanza sent over s2sout
Jan 03 17:40:34 s2sin53cbc60    debug   Received[s2sin]: <iq id='push' type='result' to='user@yax.im' from='pubsub.chatsecure.org'>
```

**Comment by ge0rg** [comment]
This is the full IQ-set that's sent from yax.im to ChatSecure's server (token and node anonymized):

```xml
<iq type="set" to="pubsub.chatsecure.org" from="user@yax.im" id="push">
  <pubsub xmlns="http://jabber.org/protocol/pubsub">
    <publish node="A9BFF3A9-276C-46E2-****-************">
      <item>
        <notification xmlns="urn:xmpp:push:0">
          <x xmlns="jabber:x:data" type="form">
            <field type="hidden" var="FORM_TYPE">
              <value>urn:xmpp:push:summary</value>
            </field>
            <field type="text-single" var="message-count">
              <value>6</value>
            </field>
            <field type="text-single" var="pending-subscription-count"/>
            <field type="jid-single" var="last-message-sender"/>
            <field type="text-single" var="last-message-body"/>
          </x>
        </notification>
      </item>
    </publish>
    <publish-options>
      <x xmlns="jabber:x:data">
        <field var="FORM_TYPE">
          <value>http://jabber.org/protocol/pubsub#publish-options</value>
        </field>
        <field var="token">
          <value>b86cd972bcfb466f************************</value>
        </field>
        <field var="endpoint">
          <value>https://push.chatsecure.org/api/v1/messages/</value>
        </field>
      </x>
    </publish-options>
  </pubsub>
</iq>
```

**Comment by chrisballinger** [comment]
I'm seeing this as well, not sure why it stopped working because we've made no changes in a long time. A quick look at our error logs shows no clear culprit, so I'll have to take a closer look. Let me know if you notice it working again.

Does the manual "Knock" feature work between two ChatSecure iOS clients?

**Comment by chrisballinger** [comment]
Okay, so I tested again this afternoon and was able to get it to work using both Knock as well as XEP-0357 style pushes.

This might be OS throttling based on our use of content-available..

**Comment by ge0rg** [comment]
Upgraded both of my devices to beta 58. XEP-0357 notifications still don't arrive at the device.

The iPhone was attached to the charger all the time, low-power-mode is disabled.

Not quite sure how to trigger the "Knock" feature, IIRC it should appear instead of "Send" in the chat view for offline contacts? My secondary iOS device/account is offline, but all I see is a greyed out "Send" button.

**Comment by chrisballinger** [comment]
In order for Knock to work between iOS clients, you need to establish an OTR session once in a while to exchange tokens, make sure you set it to "OTR" or "OMEMO & OTR", at least temporarily.

**Comment by ge0rg** [comment]
Just a short update: push still doesn't work on 4.0.1, and I wasn't able to see the "Knock" button despite having OTR configured on both devices.

**Comment by chrisballinger** [comment]
@ge0rg 
1. If you go to ChatSecure's Settings do you see "Enable Push" as an option below the accounts? 
2. If you go to iOS System Settings -> ChatSecure, are "Notifications" and "Background App Refresh" both turned on?

**Comment by ge0rg** [comment]
1. no, neither on the iPod nor on the iPhone6
2. yes and yes

**Comment by afriedmanGlacier** [comment]
I've been testing this same issue a bit of push notifications working sporadically, and @chrisballinger I think you are right about it being due to use of content-available, making it a silent notification. I am using ChatSecure-Push-Server, and I modified the notification in push/messenger/views.py to remove the content_available=True argument, and I added the message text "New Message" just to see what would happen and, at least with initial testing, I now receive the push notification every time. 

In order to avoid silent notifications and get the notifications every time, and maintain security, maybe need to use the "Modifying Payload of Remote Notifications" feature in iOS? This way, send a generic message that allows the app time to connect to the server, retrieve the new message, and change the text of the notification to that of the new message. See https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/ModifyingNotifications.html#//apple_ref/doc/uid/TP40008194-CH16-SW

I haven't gone down the road of trying this yet (and won't have time this week), and I'm guessing you guys are way ahead of me, but just in case it helps.

**Comment by chrisballinger** [comment]
Unfortunately with the first method you would see a lot of generic "New Message" messages on the screen, because XMPP servers are a little spammy at the moment (it will push for things like typing notifications). 

I think you might be on to something with the mutable notification thing, but it depends on how much time we'd get in the background. I didn't realize that you could make background network requests with mutable notifications, I was under the impression that you had to include the payload and decrypt it offline. Another snag is that it only works on iOS 10.

When we originally designed the push stuff, we considered leaving in an option for a generic "New Message" push, so you could attempt to force through background throttling, but it added too much complexity for the sending logic.

**Comment by therob84** [comment]
I am not sure if this is the right topic, maybe you can point me elsewhere if not.

In the current version (4.0.2) how is the 
a) delivery of offline-messages handled and 
b) in which way is ChatSecure informed about new messages while the app is not active in the foreground?

Background of the question: I use Conversations @Android with a friend using ChatSecure@iOS. I want to know, in which circumstances my messages won't be delivered at the friends iOS and if he should change any settings to improve the situation...

Thanks for answers or links...

**Comment by chrisballinger** [comment]
@therob84 

If you are using a server that supports XEP-0357, the server can trigger a "wakeup" notification that logs in the client and downloads any queued messages. This also depends on server support for offline queueing of course. The app also supports periodic background fetch, so even if you don't have push, the app may occasionally check for new messages on its own, but we can't control how frequently that occurs.

The number 1 thing you can do is to make sure your server supports XEP-0357 and offline queueing.

**Comment by davepb16** [comment]
I'm using prosody 0.10 with the cloud_notify module.  As far as I can tell, and after looking though these threads, my configuration is fine.  My certificate is from Letsencrypt and is good.

With my Android clients, using Conversations, I can watch when they come online and can see push notifications go to those clients.

With Chatsecure-ios, using a brand new user account or an existing account, I see none of this push activity in the prosody debug logs.  It's as if enabling "push notifications" on IOS doesn't do anything.
After 5 minutes, smacks destroys the session due to inactivity from the Chatsecure Client and I can no longer send any messages to the device.

Is there something I should be looking out for to remedy this?   

Thanks.

**Comment by therob84** [comment]
@Chrisballinger - thanks for fast feedback!
Since we both using jabber.de XEP-0357 seems to be enabled.
But I can't see any "received" marker nor "read" marker in conversations (thus I cannot check if messages are received and only ignored, or if the case of @davepb16 is valid). OMEMO is working!

So there is nothing I should tell the iOS-friend to change...?!

**Comment by chrisballinger** [comment]
@davepb16 Does your cert include the client bit as well as server bit? See here: https://github.com/ChatSecure/ChatSecure-iOS/issues/557#issuecomment-247479188

You should also try enabling `mod_offline` and `mod_smacks_offline`. 

@therob84 We don't support read receipts yet, only delivery receipts. 

We should probably do a better job in the UI indicating whether or not your server works with push.

**Comment by davepb16** [comment]
@chrisballinger 

Certificate has both bits set...
            "X509v3 Extended Key Usage: 
            TLS Web Server Authentication, TLS Web Client Authentication"

I have both smacks and smacks_offline set. 

When a **Conversations client** (push capable) logs in to prosody I see:
c2s1a6fd00      info    Authenticated as user1@domain.co.uk
..........
c2s1a6fd00      debug   Received[c2s]: \<presence from='user1@domain.co.uk/phone'\>
..........
c2s12a3a18 debug   **Received**[c2s]: \<iq id='12345abcd' type='set' to='push.siacs.eu' from='user1@domain.co.uk/phone'\>
..........
s2sout195ae60   debug   **sending**: \<iq type='set' to='push.siacs.eu' from='user1@domain.co.uk/phone' id='12345abcd'\>

With a **Chatsecure-ios** client I see:
c2s1a6fd00      info    Authenticated as user2@domain.co.uk
..........
c2s1a6fd00      debug   Received[c2s]: \<presence from='user2@domain.co.uk/Chatsecure12345'\>

and nothing more about push notifications for the ios client.

There was an anomaly in my dns SRV record that I've just amended.  A sub domain that wasn't included in my certificate.

I'm sure I had push working on Chatsecure-ios at one stage.

Tomorrow when I get hold of an iPhone I will set up a new user on prosody and use OTR only to see if it makes a difference.

Thanks for the advice.

**Comment by gerg5c42g542g2c54g52c** [comment]
Just wanted to report that I tested most of the servers that support Push notifications as listed here: https://gultsch.de/compliance_ranked.html with the newest version of ChatSecure.
They worked only on one server, talker.to, every other server that I tested didn't notify of messages.

**Comment by davepb16** [comment]
Reporting back....

With prosody 0.10, Apple push notifications don't work.  Tried new user account, OTR and OMEMO.  No sign of push messages in log file. SSL certificate has both bits set and SRV records seem fine.

It also appears that on the iPhone, IOS 10.2.1, when the screen turns off, wifi gets disabled and Chatsecure fails to initiate another connection over the mobile data.  It also fails to initiate a connection when the screen/wifi turns on.  I have to open up the app to get Chatsecure to connect to the prosody server.

If you need any logs let me know.

**Comment by chrisballinger** [comment]
What's curious is that it definitely works for me on my Prosody setup and v4.0.2 of the app. Below are all of the modules I have enabled.

I believe that these are all required for it to work reliably. Also make sure the "offline" is not disabled (it should be auto loaded).
```lua
                "cloud_notify";
                "pinger";
                "smacks";
                "smacks_offline";
```

Full:
```lua
modules_enabled = {

        -- Generally required
                "roster"; -- Allow users to have a roster. Recommended ;)
                "saslauth"; -- Authentication for clients and servers. Recommended if you want to log in.
                "tls"; -- Add support for secure TLS on c2s/s2s connections
                "dialback"; -- s2s dialback support
                "disco"; -- Service discovery

        -- Not essential, but recommended
                "private"; -- Private XML storage (for room bookmarks, etc.)
                "vcard"; -- Allow users to set vCards

        -- These are commented by default as they have a performance impact
                --"privacy"; -- Support privacy lists
                --"compression"; -- Stream compression

        -- Nice to have
                "version"; -- Replies to server version requests
                "uptime"; -- Report how long server has been running
                "time"; -- Let others know the time here on this server
                "ping"; -- Replies to XMPP pings with pongs
                "pep"; -- Enables users to publish their mood, activity, playing music and more
                "pubsub";
                "cloud_notify";
                "pinger";
                "smacks";
                "smacks_offline";
                "mam";
                "csi";
                "carbons";
                --"register"; -- Allow users to register on this server using a client and change passwords

        -- Admin interfaces
                --"admin_adhoc"; -- Allows administration via an XMPP client that supports ad-hoc commands
                --"admin_telnet"; -- Opens telnet console interface on localhost port 5582

        -- HTTP modules
                --"bosh"; -- Enable BOSH clients, aka "Jabber over HTTP"
                --"http_files"; -- Serve static files from a directory over HTTP
               --"http_files"; -- Serve static files from a directory over HTTP

        -- Other specific functionality
                "posix"; -- POSIX functionality, sends server to background, enables syslog, etc.
                --"groups"; -- Shared roster support
                --"announce"; -- Send announcement to all online users
                --"welcome"; -- Welcome users who register accounts
                --"watchregistrations"; -- Alert admins of registrations
                --"motd"; -- Send a message to users when they log in
                --"legacyauth"; -- Legacy authentication. Only used by some old clients and bots.
};
```

**Comment by davepb16** [comment]
@chrisballinger 

I've got all of those modules enabled apart from "carbons" and "mam".

I have taken a 5 minute snapshot of Prosody debug logs showing the iPhone client authenticating, and then getting disconnected as the mod_smacks hibernation timeout was reached (300 seconds).  Nothing is seen from it again, even if the iPhone screen is turned on.

I then added a "conversations.im" XMPP account to both Android client and iPhone client (in conjunction with my Prosody accounts) just to see if it was an issue with my Prosody setup or Certificate setup.

I noticed that when the Prosody smacks_hibernation_time expired, and I lost complete contact with that ios account, I could type into the "online" conversations.im message window and the Prosody account would come "online" again.

The smacks_hibernation_time on the "conversations.im" must be about 30 minutes as it remained "online" a lot longer than my Prosody account.

Now the smacks_hibernation_time has expired on the "conversations.im" server as both XMPP accounts now show offline and I can't wake either up, by typing or sending a message. 
I'm guessing this means that Push notifications aren't being send from the "conversations.im" server to the iPhone either.

The only thing I can do now is to try to get different SSL certificates, try ejabberd or setup Prosody again.
I'm sure Push notifications were working in Chatsecure 4.0 and my Prosody server.

:)

debug logs (identities removed):  [http://pastebin.com/Sdqv5shF](http://pastebin.com/Sdqv5shF)

**Comment by chrisballinger** [comment]
@davepb16 Wait so you're seeing absolutely NO logs on your server about push registration?. Set your Prosody log level to debug and look for `mod_cloud_notify` logs here:

https://hg.prosody.im/prosody-modules/file/tip/mod_cloud_notify/mod_cloud_notify.lua#l26

If you aren't seeing any of those logs at all then there must be something wrong with the client that's preventing it from registering.

**Comment by davepb16** [comment]
@chrisballinger 

I definitely get push registration for my Android clients.  Just none of my iPhone clients.  The iPhone clients  are on different networks, one is an iPhone 5s and one's an iPad,  both are on ios 10.0.2.1

My logs are already set to debug to try to find out the issue here.

Here's a typical log of push registration messages of one of my Android clients after it authenticates.

[https://pastebin.com/7Lee5Kct](https://pastebin.com/7Lee5Kct)

**Comment by chrisballinger** [comment]
@davepb16 Ok at least that narrows down the problem to the iOS client. The things that could cause this are:

1. "Enable Push" was skipped in the onboarding
2. The push notification permissions are disabled
3. Background app refresh is disabled (only for receiving push, not registering)
4. The [ChatSecure Push API](https://push.chatsecure.org/api/v1/) account is broken for some reason and cannot receive new tokens
5. Somehow the query to the xmpp server asking for its push capabilities is broken

From the logs it looks like maybe (5) is the case?

Also can you try testing iOS to iOS and verifying if the "Knock" feature works? In order to test you must establish an OTR session and then the other party needs to go fully offline. It will replace the Send button with a Knock button when no text is in the field. In this case you'll probably need to manually logout one side so the stream management timer doesn't waste your time.

**Comment by davepb16** [comment]
1. "Enable Push" was selected whist installing the app/account on both ipad and iphone.
2. Not sure where to check this on ios 10 but notifications for Chatsecure were enabled in the ios settings.
3. Background app refresh enabled but not sure how to check for 'registering'.
4. Would this happen to two separate Apple devices housed at different locations?
5. Agree that it looks the most likely.

I'm still not ruling out a problem at my end with prosody/cloud_notify/ssl cert.  I might try to setup a new prosody server (or maybe ejabberd) with new certificates on a different machine and see.

Are there any particular SRV DNS settings (for my domain) that Chatsecure needs to work with push and prosody?

I will try the 'knock' soon, when I get the two Apple devices together.

Thanks.

**Comment by chrisballinger** [comment]
I'm pretty sure it is a client issue because you should see cloud_notify registration in your logs. In a future version we will make it easier to see if you're registered for our push API, and if your server is advertising support.

**Comment by mt403** [comment]
I can confirm @davepb16  issue. I triggered the issue by reinstalling the app due to omemo issues. After that push never started working again. Issue #634  should give a ruff estimate of the time frame when it got broken.

**Comment by chrisballinger** [comment]
@mt403 I'm going to work on a better server capabilities UI to help diagnose this issue. I have a feeling that the push API account is possibly in a bad state so the app thinks that push is unavailable.

Deleting and reinstalling the app resets push permissions.. so we might be on to something.

**Comment by mt403** [comment]
@chrisballinger Reinstalling the app does not enable push atm. I've tried all betas since the 4.0 release. I always select the enable push option after account creation. The ios settings dialog for chatsecure only lists background something, should there be more options ?

**Comment by chrisballinger** [comment]
Ah yes that's the problem. For some reason we must not be re-prompting for push permissions, so it's never able to finish. This is what it should look like:

![simulator screen shot feb 9 2017 9 56 15 am](https://cloud.githubusercontent.com/assets/518687/22796115/03ef119e-eeae-11e6-9498-d3b85f3a2ee4.png)
![simulator screen shot feb 9 2017 9 56 17 am](https://cloud.githubusercontent.com/assets/518687/22796120/07c206aa-eeae-11e6-9818-88f138306263.png)

**Comment by mt403** [comment]
@chrisballinger Settings look the same for me. I was not specific enough, i always did reinstall the app in my tests. Shouldn't re asking for push permissions be fine in that case ? Because i thought removing the app would als remove all settings.

**Comment by chrisballinger** [comment]
I believe the behavior differs in different versions of iOS, but generally when you hit "Enable Push" in the onboarding it should prompt you with the system message for allowing push notifications. And then if that doesn't work for some reason, we will show an "Enable Push" row in the Settings view of our app.

I'm going to focus on https://github.com/ChatSecure/ChatSecure-iOS/issues/684 because I think it will help us debug the underlying cause. Are you signed up for the TestFlight beta?

**Comment by mt403** [comment]
I can't recall any iOS prompts, only the cs onboarding. I also dont have a
push option inside the app. Its supposed to be in settings accessed by the
cog wheel right ?

Yes I'm already a beta tester. Testing with an iPad 4 with the latest iOS
available.

**Comment by therob84** [comment]
> The app also supports periodic background fetch, so even if you don't have push, the app may occasionally check for new messages on its own, but we can't control how frequently that occurs.
The number 1 thing you can do is to make sure your server supports XEP-0357 and offline queueing.

I shortly want to come back to my question: "_What can I tell to my non-techy friend about checking/changing any settings in ChatSecure or iOS itself in order to receive also my messages after she did not used ChatSecure actively for one week?_" (both jabber.de with XEP-0357 enabled, I am using Conversations).

Since I never saw all the settings of ChatSecure or iOS I don't know, if any critical toggles exist, in order to enable relieable delivery of messages to my ChatSecure-Friend (just saw your screenshots above). At the moment this is unfortunatley not the case - as she did not receive any message after our initially successfull OMEMO-conversation (at least there was no anwser and no "delivery" markers in Conversations without these problems with Conversations-Friends).

> "Enable Push" was skipped in the onboarding

Is this anything my friend should have manually set during installation of ChatSecure? How to change afterwards?

> The app also supports periodic background fetch, so even if you don't have push, the app may occasionally check for new messages on its own, but we can't control how frequently that occurs.

What does it mean that you can't control it? Its an iOS thing? Will it happen once a minute or once a year? Mysterious statement...

...Never thought that I would have to deal with iOS-problems just to bring XMPP+OMEMO also to my friends with iOS :-/ While it is running very easy with Conversations...
I hope the best for some improvements in this topic within the next updates and nevertheless I am still very thankful about the great progress @chrisballinger is acheiving with OMEMO-Integration.

**Comment by chrisballinger** [comment]
@therob84 I am working on a [server features browser](https://github.com/ChatSecure/ChatSecure-iOS/issues/684) that will also include some helpful info for diagnosing and fixing issues with push. Ideally everything should "just work" if your friend did "Enable Push" in the onboarding, but there may  be edge cases that are hard to diagnose remotely until I get this view into the app.

> What does it mean that you can't control it? Its an iOS thing? Will it happen once a minute or once a year? Mysterious statement...

[Here's an overview](https://www.objc.io/issues/5-ios7/multitasking/) of the background fetch APIs and their limitations. We use background fetch and "content-available" push notifications, which are both not 100% reliable and throttled by the OS. Conversations doesn't have to deal with any of this because you can still disable doze and use a long lived socket.

**Comment by therob84** [comment]
> We use background fetch and "content-available" push notifications, which are both not 100% reliable and throttled by the OS.

Thanks for the link! Also this means, that these features could also not be available with iOS at versions prior to 7. Which feature ChatSecure uses in this case for "Push" etc. (which I think is true for my friends device)?

**Comment by chrisballinger** [comment]
@therob84 We use "content-available" push notifications which have existed since iOS 7. However the oldest version of iOS we support is 8.0, so your friends device must have 8 or higher if they were also doing OMEMO.

**Comment by afriedmanGlacier** [comment]
@chrisballinger just curious, are you planning to move away from "content-available" as part of resolving this ticket since the operating system seems to be throttling silent push messages?

**Comment by chrisballinger** [comment]
@afriedmanGlacier Yes, but it's not an easy problem to solve. Here are the options:

1. Add VoIP functionality to the app so we can abuse high priority pushes. (not likely to happen any time soon)
2. Use iOS 10-only push notifications that allow you to do a similar high priority fetch... but it needs to map 1:1 to something displayed on the screen, and you get a similar 30 second timeout. I also believe it requires rewriting large pieces of the app so it works within an app extension. XMPP servers often send us pushes for things like typing notifications, which are actually beneficial because it will usually wake up the client before a message is sent. If these new notifications are used, the typing notifications will need to be displayed to the user with a static string and can be noisy. Also if a push doesnt match to a message or anything at all, it still needs to show a static message like "New Message", even if that isn't necessarily true.

**Comment by davepb16** [comment]
@chrisballinger 

After ChatSecure registers with the Apple Push Notification server does it pass a device token back to the xmpp (prosody) server?

If I look in my prosody mysql database I can see mod_cloud_notify has these tokens for all of my Android clients but none for ChatSecure ios clients.  Looking though my debugs logs, ChatSecure-ios doesn't seem to return them either.

**Comment by chrisballinger** [comment]
@davepb16 [Here is a description](https://github.com/ChatSecure/ChatSecure-Push-Server/tree/master/docs/v3#level-2) of how "level 2" push works. Instead of using the device token directly, we have users fetch time-limited tokens to reduce identifiable metadata. The flow goes like this:

1. User enables push notifications, and Device token is retrieved.
2. If push is enabled and we have a Device token, Push API account is created (POST `push.chatsecure.org/api/v1/account`)
3. Device token sent to push API (POST `push.chatsecure.org/api/v1/device/apns`)
4. Knock tokens fetched from push API (GET `push.chatsecure.org/api/v1/tokens`)
5. Every connect, a fresh knock token, Push API endpoint and pubsub endpoint (`pubsub.chatsecure.org`) is sent to XMPP server
6. When XMPP server wants to push, it does a POST to `push.chatsecure.org/api/v1/messages` with the knock token.
7. Push API server receives Knock token and looks up Device token.
7. Push API server uses Device token to send background wakeup push via APNS.

If the Knock tokens / Push API endpoint aren't being sent to the XMPP server, then there must be a problem with the 1st thru 4th step. I am creating a new view to help people debug these problems here: https://github.com/ChatSecure/ChatSecure-iOS/issues/684#issuecomment-280112514

**Comment by davepb16** [comment]
@chrisballinger 

Thanks.  I remember this working initially in ChatSecure 4.0.0, using prosody 0.10.
I could type/send message to an ios client (using OTR) and wake it up.  I never kept debug logs though.

Can't remember when it stopped working but I'm guessing after ChatSecure 4.0.1.

As a point of interest, if I create/use xmpp accounts for "Conversations.im" (ejabberd server), I have some success waking up ios ChatSecure client from the Android conversations app by typing a message to it.  
Part of the push notification system must still be working.

Looking forward to [#684](https://github.com/ChatSecure/ChatSecure-iOS/issues/684#issuecomment-280112514)

**Comment by chrisballinger** [comment]
@davepb16 Are you saying if you use ChatSecure to connect to an account created on the Conversations.im server, that push notifications to ChatSecure work?

**Comment by davepb16** [comment]
About 90% of the time yes.

I can start to type from Android "conversations.im" account to ChatSecure-ios "conversations.im" account and wake it up.  It also wakes up any prosody accounts that are also on ChatSecure-ios client.

I only have issues with Prosody servers and Chatsecure-ios.  I haven't tried many other XMPP servers though.  

I did try dukgo.com and push didn't work.  I think dukgo.com is using prosody 0.9 though.
conversations.im, which works, uses ejabberd server.

I know other people in this thread tried more servers.

**Comment by chrisballinger** [comment]
dukgo doesn't have push support and likely never will. My personal server is running Prosody 0.9.x with mod_cloud_notify and works very reliably with push.

**Comment by davepb16** [comment]
ok, I think I will have to wipe my server and start again from scratch.  Maybe I will try with prosody 0.9.x instead of 0.10.  Or maybe even ejabberd.

Thanks for your help.

**Comment by chrisballinger** [comment]
@davepb16 Are you on the beta? Instead of wiping your server now, wait until next week, I will push a beta with #684 because maybe that can diagnose it.

**Comment by davepb16** [comment]
Not on beta, no.  I will wait before I wipe the server.  Not in a rush to do it :)

**Comment by davepb16** [comment]
@chrisballinger 

Problem fixed...

I setup Prosody 0.9.12 using the same Letsencrypt certificates, almost identical configuration files and a new mysql prosody database and push now works as expected.  

My logs are now full of push.chatsecure.org comments and "pubsub.chatsecure.org<E123456AB-7EE4-4220-8083-123456ABCDEF" now fills my mysql database for mod_cloud_notify entries.

I guess this issue is with prosody 0.10 running on ArchLinux.

If needed I can setup a broken prosody 0.10 again for testing purposes.

Thanks for your help.

**Comment by chrisballinger** [comment]
@davepb16 Oh that's awesome! It seems like something in 0.10 was preventing mod_cloud_notify from loading correctly?

**Comment by silberzwiebel** [comment]
@davepb16 It would be cool, if you could setup a broken prosody 0.10 to see where the fault is on prosody's side. I'm also running a prosody 0.10 server and seeing this issue with a friend who I'm trying to convince to use XMPP. Only seeing messages after opening the app is a show-stopper, though.
I'd rather see a potential issue fixed in prosody 0.10 than downgrade to 0.9.12.
Many thanks!

**Comment by davepb16** [comment]
@silberzwiebel 

I'm happy to do that.  Busy this week but will hopefully have some time next week to do it.

We should open a separate thread for it so as to not fill this one up.

Also, waiting for @chrisballinger to release a new version of Chatsecure-ios that will give information on the XMPP server capabilities and PUSH notification issues.

My email is in my profile if you need to contact me.

**Comment by tmolitor-stud-tu** [comment]
@davepb16 The problem with prosody 0.10 is this: #712 
Current versions of Chatsecure don't provide an id on iqs and prosody 0.10 is stricter in not allowing such iqs.

**Comment by tmolitor-stud-tu** [comment]
@chrisballinger I solved the `id` problem (see #712) by manually reissuing the enable stanza via sendxmpp commandline client (and specifying an `id` this time). Later I also disabled the `id` check in prosody entirely.

Now my server sends out push messages as expected but nothing happens on the client :(

Interestingly I can send push messages via https://push.chatsecure.org/api/v1/messages/ using the token registered with my xmpp server and those push messages are displayed and can be used to wake up chatsecure by opening them (this doesn't happen automatically).

But pushes send from my xmpp server don't seem to make it to the device.
This is an example of the push stanza (anonymized, if you need the full stanza just ask):
```
<iq type="set" to="pubsub.chatsecure.org" from="user@example.org" id="pubsub.chatsecure.org&lt;DEADBEEF-DEAD-BEEF-DEAD-BEEFDEADBEEF">
    <pubsub xmlns="http://jabber.org/protocol/pubsub">
    <publish node="DEADBEEF-DEAD-BEEF-DEAD-BEEFDEADBEEF">
        <item>
            <notification xmlns="urn:xmpp:push:0">
                <x xmlns="jabber:x:data" type="form">
                    <field type="hidden" var="FORM_TYPE">
                        <value>urn:xmpp:push:summary</value>
                    </field>
                    <field type="text-single" var="message-count">
                        <value>1</value>
                    </field>
                    <field type="text-single" var="pending-subscription-count"/>
                    <field type="jid-single" var="last-message-sender"/>
                    <field type="text-single" var="last-message-body"/>
                </x>
            </notification>
        </item>
    </publish>
    <publish-options>
        <x xmlns="jabber:x:data">
            <field var="FORM_TYPE">
                <value>http://jabber.org/protocol/pubsub#publish-options</value>
            </field>
            <field var="token">
                <value>anonymized_token_value</value>
            </field>
            <field var="endpoint">
                <value>https://push.chatsecure.org/api/v1/messages/</value>
            </field>
        </x>
    </publish-options>
</pubsub>
</iq>
```
**Your sever's answer:**
`<iq from="pubsub.chatsecure.org" to="user@example.org" id="pubsub.chatsecure.org&lt;DEADBEEF-DEAD-BEEF-DEAD-BEEFDEADBEEF" type="result"/>`

**If you need some more info, I'd be glad to help.**

**Comment by tmolitor-stud-tu** [comment]
@chrisballinger @iNPUTmice Wouldn't it be possible to extend XEP-0357 to allow the server to specify a priority for every push message?
This way the server could trigger low priority (silent) pushes for typing notifications and other thanzas containing no message body and high priority pushes for message stanzas having a body.

Wouldn't that solve most of the issues described here: https://github.com/ChatSecure/ChatSecure-iOS/issues/637#issuecomment-279450445

**Comment by tmolitor-stud-tu** [comment]
@chrisballinger Could you contact me via xmpp at tmolitor@datenknoten.me to discuss those XEP changes further?
I already had a longer discussion with Holger about this issue and we had several ideas to improve the situation that I want to discuss with you.
I don't have much experience regarding iOS and your experience here is invaluable to sort out those ideas that actually could work.

**Comment by chrisballinger** [comment]
@tmolitor-stud-tu We already forward pushes for typing notifications. All notifications are silent and just trigger the app to wake up for ~30 seconds, any notifications you see are generated locally on the device. Btw, please don't send `last-message-sender` and `last-message-body` because we don't want that information and purposefully discard these values

I have a new beta that will be sent out soon that will fix the iq issue, among other things. Would you mind sending me an email so I can invite you to the next TestFlight release?

**Comment by iNPUTmice** [comment]
Why send a push notification at all for unimportant things like Chat States?
Anything that will trigger a notification on the user's phone should be
pushed with the highest possible priority and anything else just shouldn't
be pushed at all. It should probably be tied to CSI. If CSI things it's
worthy to be send, send a push. If not don't.
I can't think of an example where I would prefer a low priority push over
no push at all.
But maybe I'm missing something?

**Comment by tmolitor-stud-tu** [comment]
@iNPUTmice I think you are missing `iq` stanzas. Those should not be displayed as notification to the user but wake up the device so it can respond to them.

**Comment by tmolitor-stud-tu** [comment]
@chrisballinger do you need my private email address or my apple id to do this?

**Comment by weiss** [comment]
> anything else just shouldn't be pushed at all.

If we want to maintain a server-side session as long as the client is reachable we must be able to ping the client (e.g. after server restarts or after some idle timeout).  The server must also be able to flush the CSI queue.  If there's no way to make "silent notifications" reliable on iOS, I think we must give up the idea of keeping an online session for iOS clients.

**Comment by weiss** [comment]
> All notifications are silent and just trigger the app to wake up for ~30 seconds

I thought the point of this issue was that silent APNS notifications don't seem to be reliable?  Is your plan to stick to them nevertheless?

**Comment by chrisballinger** [comment]
@iNPUTmice Sending a notification for chat state change helps because it has the side effect of waking up the other client before you've finished writing your message. There is no concept of priority for our pushes in the current design, nor can there be. They are all "low priority" background pushes without any content that simply wake up the device for ~30 seconds. We will never include plaintext content in our pushes.

Your server can choose to not send pushes for unimportant events like chat states, but the usability is slightly worse because of the slight lag between push -> client awake.

@tmolitor-stud-tu  Our pubsub server will trigger a push whenever you send it a valid push iq. It's up to your server to decide when to trigger these.

Any email will do, you connect your Apple ID later during the signup process.

**Comment by tmolitor-stud-tu** [comment]
@iNPUTmice @weiss Wouldn't be muc invites also a thing for which you want silent pushes instead of no pushes at all?

@chrisballinger As far as I understand the phone displays the text sent with the APNS notification and to not disclose the message to Apple we have to use silent notifications (which aren't reliable).
Sending high priority pushes with a placeholder text is problematic and those messages would have to be replaced by the actual message content received via XMPP after the device/app wakes up.

**Comment by iNPUTmice** [comment]
> @iNPUTmice I think you are missing `iq` stanzas. Those should not be displayed as notification to the user but wake up the device so it can respond to them.

Well IQ+push is kinda uncharted territory for now. My guess is that the server might either a) try to answer them (caps caching) or b) send an automated error response for like really stupid unnecessary things some bad behaving clients might send (Thinking of client to client ping for example.) or c) try to deliver them as soon as possible. Think jingle file transfer or voice for example. Even with IQs in mind I have a hard time coming up with a use case for 'hey here is this push but answer that when ever'. Most IQ are fairly time sensitive as the sending side probably has timeouts in place as well.

@weiss I was just responding to a adding priorities to XEP-0357. Even if the situation on iOS is that we can't reliably deliver notifications I'm not sure how adding priorities to the XEP changes that situation.

**Comment by tmolitor-stud-tu** [comment]
@chrisballinger No, I don't want you to send the message content as plaintext through apples servers. That would be very bad!

**Comment by iNPUTmice** [comment]
> @iNPUTmice @weiss Wouldn't be muc invites also a thing for which you want silent pushes instead of no pushes at all?

I would prefer a high priority push for that. Because only after receiving the invite I can join the MUC, catch up with MAM and actually check for messages that may or may not trigger notifications.

Or if your client goes the do not auto accept invites route (fine as well) you might want to trigger a notification asking the user to accept the invitation.

**Comment by chrisballinger** [comment]
@weiss Silent notifications are throttled by the OS and won't work if you have low power enabled, you've force closed the app, or you've restarted the device. From my experience, they work well enough for me to receive messages most of the time.

In the future we can (ab)use iOS 10 mutable notifications, as they are more reliable than `content-available`, but it requires a 1:1 mapping between a push and content to display. If you fail to fetch content, it must display some sort of default message to the user. This would require sending additional data down the pipe so we know if a push corresponds to a typing notification, or filtering out typing notifications altogether. Snapchat sends pushes for typing notifications, so maybe it's not terrible to show those.

Adding iOS 10 mutable notifications isn't easy, as it requires a MAJOR refactor of the app to allow for a multi-process design. See [UNNotificationServiceExtension](https://developer.apple.com/reference/usernotifications/unnotificationserviceextension). Since I plan to start working on a macOS port over the summer, there will be a lot of refactoring during this period to allow for multi process sandboxing.

**Comment by tmolitor-stud-tu** [comment]
@iNPUTmice On iOS silent pushes aren't reliable and high priority pushes always display a notification containing the text specified by the push server.
With the current API it isn't possible for iOS developers to have reliable pushes AND auto accept invites to check MAM etc.

Having a way in the XEP to distinguish pushes which contain actual messages instead of some invite or iq or something else, would be good. This way high priority pushes could be sent for those messages and silent ones for everything else.

This way message could still be delivered even when the device is in power saving mode or the app was force closed.

**Comment by iNPUTmice** [comment]
@tmolitor-stud-tu so just to clarify you want to send a high priority push messages for things that will **definitively** trigger a notification and low priority push messages for something that **might** trigger notifications?

**Comment by tmolitor-stud-tu** [comment]
@iNPUTmice Exactly :) Otherwise (e.g. when everything is pushed as high priority) there could be useless notifications floating around irritating users and spamming the lockscreen.

**Comment by tmolitor-stud-tu** [comment]
@iNPUTmice Ideally there would be some sort of api to specify which stanzas the client wants as high priority push and which ones as low priority....maybe along with an identifier to further identify certain stanzas.
This could be used by the push server to display different messages to the user in the high priority case like "New Message", "Incoming MUC invite" etc.

**Comment by weiss** [comment]
> In the future we can (ab)use iOS 10 mutable notifications

Couldn't you use "normal" (non-silent) high priority notifications already, just that this wouldn't allow you to edit the notification text?  I mean, couldn't you send a generic "new message" text without refactoring ChatSecure and depending on iOS 10?

From my experience, and from various user comments I've seen in this issue and elsewhere, the current solution does feel quite unreliable in practice.  Maybe users would be happy to live with generic "new message" notifications if this makes things reliable?  (This is also what the Matrix client "Riot" seems to be doing when E2EE is enabled, by the way.  Not sure how other messengers solve the issue on iOS ...)

**Comment by iNPUTmice** [comment]
@tmolitor-stud-tu I'd design it the other way round. Instead the client telling the server for what types it wants a HIGH or LOW priority I'd let the server tell the app server more about the type and let the app server decide.
So a server to app server push would contain a type=readable-message, or type=invite or something like that.

**Comment by chrisballinger** [comment]
We can send a high priority push that also fetches in the background, but it must be accompanied by text/sound/badge and this content cannot be modified at runtime. That allows for a push that says something generic like "Incoming Message", which then does some background stuff, and then generates local notifications as it does currently.


> The notification’s priority. Provide one of the following values:
>
> 10 The push message is sent immediately.
> The push notification must trigger an alert, sound, or badge on the device. It is an error to use this 
> priority for a push that contains only the content-available key.
> 5 The push message is sent at a time that conserves power on the device receiving it.

**Comment by tmolitor-stud-tu** [comment]
@iNPUTmice Well, this would also be possible. But we need an exhaustive list of types to help the app server decide what message it will show for high priority pushes.

Having those pre-defined types is not as flexible and future proof as having the client specify which stanzas it wants and assigning a type to them.

But defining an api for those "stanza filtering" in the XEP is a big change whereas your proposal is a bit smaller and easier to implement on the xmpp server side. While making the app server side not much harder to implement.

I'd go with your proposal as far as I'm concerned :)

**Comment by tmolitor-stud-tu** [comment]
@chrisballinger Is it already possible to wake up the app by using https://push.chatsecure.org/api/v1/messages/ directly (with a generic alert text like "New Message")?

If yes, I could hack my server implementation to use this api endpoint directly for high priority messages without you changing the app at all.

Would be a **HUGHE** improvement for me!

**Comment by afriedmanGlacier** [comment]
I'm late to the party, but we are doing what @weiss and @tmolitor-stud-tu are suggesting in sending a generic "New Message." We decided this was a better solution for us in that we always want to receive the push notification and we weren't getting them all as silent pushes. Using eJabber and a modified version of @weiss mod_push along with ChatSecure-Push-Server and RubDub. mod_push is changed so that we aren't getting things like typing notifications and non-messages. Changed the push notification in CS Push Server to be non-silent "New Message." I'm not sure our solution is the best for everyone, but it works well for our needs and we get all of our pushes. The thing I'm not sure about is how our current version plays with MUC because we haven't been using it yet. Just in case that helps anyone :)

**Comment by chrisballinger** [comment]
@tmolitor-stud-tu We can change the behavior of our [pubsub module](https://github.com/ChatSecure/RubDub) and push API server to allow for this without needing to modify the clients at all.

I think that some types should be defined so that my pubsub node translate or filter the requests before passing them off to our push API server. Some ideas and how they could map to hard coded localizable strings:

* `message`: "New Message"
* `typing`: "Someone is typing..." (might have to filter these out entirely though to prevent UI clutter)
* `invite`: "Incoming Request"

**Comment by tmolitor-stud-tu** [comment]
@chrisballinger This would be really awesome! :)
I'll create a list of stanza types tomorrow.

**Comment by tmolitor-stud-tu** [comment]
@chrisballinger @iNPUTmice @weiss 
I would propose the following notification types:

- `presence` (all presence stanzas)
- `important_presence` (for joining mucs, not sure how to define what muc presences are important and what are not, though, as I'm not that much into mucs)
- `invite` (muc invites)
- `chatstate` (all chatstates, could be minimized to only typing notifications)
- `readable_message` (message stanzas having a body)
- `query` (iq stanzas)
- `message` (all other message stanzas not covered above)

Any thoughts on this?

**Comment by weiss** [comment]
> I would propose the following notification types:

I think I'd avoid predefining notification types.

One idea might be to let clients add an `include_payload` (or whatever) attribute to the `<enable/>` request, which is `none` by default but can be set to `full` or `stripped` instead.  If it's set to `full`, the published `<notification/>`s would include a `<payload/>` child which holds the complete stanza.  With `stripped`, it would strip all stanza contents except for the stanza's name, its `type` attribute (if any), and the direct child elements with their `xmlns` attributes (if any).  Or something like that.

Maybe the (underspecified) `urn:xmpp:push:summary` thing could then be removed from the XEP.

Anyway, these things should probably be discussed on the [standards list][1] instead.

[1]: https://mail.jabber.org/mailman/listinfo/standards

**Comment by chrisballinger** [comment]
@weiss I don't think including the full stanza is a good idea for privacy reasons because the pubsub server is usually run by the app developers, not your XMPP server operator. You can then end up leaking content and metadata in two places instead of (mostly) just one. I 100% do not support including full stanzas, and I refuse to implement any functionality that relies on parsing them. I don't even feel comfortable with the current `summary` functionality that allows people to include the sender and message body (I also ignore this content). 

Allowing servers to send our pubsub bridge unwanted `full` stanzas is dangerous and creates a central point of failure. In my opinion it defeats the whole point of having a decentralized network. Servers operators might think it's required to include this data, or modules might have bad defaults. The ONLY functionality we need is a tap on the shoulder to wake the client up, and then do a normal connection.

**Comment by tmolitor-stud-tu** [comment]
@chrisballinger Holger proposed this setting to be entirely configurable by the client. Servers sending full stanzas when not configured to do so would violate the standard and servers can do this already, despite of any XEP changes.

There are some valid scenarios for inclusion of full stanzas.
For example a server monitoring pager used by a company which operates the XMPP server **AND** the app server as well as their own app.
_This doesn't even mean that the app server does send out the whole unencrypted stanza through GCM or APNS._

However Holger's proposal mainly was to use **stripped** stanzas instead of hardcoded stanza types.
This way no content would be leaked, but the xep and implementation thereof would be much more flexible:
> With stripped, it would strip all stanza contents except for the stanza's name, its type attribute (if any), and the direct child elements with their xmlns attributes (if any).

And last but not least Holger proposed to make the current behaviour to not send any (useful) metadata at all the default:
> One idea might be to let clients add an include_payload (or whatever) attribute to the <enable/> request, which is none by default but can be set to full or stripped instead.

I like his proposal and I think it is the right way to go (and removing the underspecified `urn:xmpp:push:summary` thing was something I thought about as well).

**Comment by weiss** [comment]
The current revision of the XEP specifies how to include message body contents with the push notification.  It doesn't specify under what circumstances the server might or should do that.  *This* is something I'd definitely remove this from the XEP.

As @tmolitor-stud-tu said, what I suggested was making this configurable per client instead, which IMO is totally harmless.  Just don't let ChatSecure enable this behavior.  I suggested having the *option* because I can imagine non-APNS scenarios where including full stanzas makes sense.  E.g., an Ubuntu Phone app cannot reconnect (to fetch stanzas) while in the background, and you might have a trusted app server there (you're not forced to use central infrastructure like on iOS or Android).  Or you might be using XEP-0357 in a totally different context, such as IoT.

As @tmolitor-stud-tu also pointed out, my main suggestion was the option to include _stripped_ stanzas with notifications (instead of specifying a predefined "notification type").  So for a chat message 

    
    <message type='chat'><body/></message>

    
    <message type='headline'><event xmlns='http://jabber.org/protocol/pubsub#event'></message>

    
    <iq type='set'><jingle xmlns='urn:xmpp:jingle:1'/></iq>

> The ONLY functionality we need is a tap on the shoulder to wake the client up, and then do a normal connection.

Initially I was really hoping this would be true on all relevant mobile platforms (which might not include things like Ubuntu Phone), but this discussion was motivated by the fact that silent notifications don't seem to be reliable on iOS, no? :-/

**Comment by chrisballinger** [comment]
> As @tmolitor-stud-tu said, what I suggested was making this configurable per client instead, which IMO is totally harmless. Just don't let ChatSecure enable this behavior.

I see, basically allow clients to tell the server how to behave, and not leave it up to the settings on the server. My nightmare scenario was that all of a sudden I get a bunch of leaky full stanzas because of poorly configured servers. There are valid cases where, if you own the whole infrastructure, you don't mind that stuff is leaking... but I don't want those cases to ever be considered default behavior. I just don't like that I can't really control what is sent to `pubsub.chatsecure.org`, so I want to minimize that as much as possible.

My hesitation with stripped stanzas is how do we know what is "safe" to include? If a client sends a more complex stanza than the strip algorithm can't handle, how do we ensure that the result doesn't leak unexpected content? Even if we can agree on a stripping algorithm, it will still leak more metadata to me than I want compared to boiling it down to just `message` and `invite` types. It also adds complexity to each 357 pubsub implementation to figure out how to parse different kinds of stanzas.

**Comment by tmolitor-stud-tu** [comment]
@chris I don't think that stripping stanzas would be overly complex.
I guess an implementation in prosody would take about 10-15 lines of code.

Yes, this is a bit more metadata than just `invite` or `message`, but in my opinion the stripped stanzas contain nothing sensitive.
Only having direct child tag names and namespaces as well as tag name and namespace of the stanza itself really isn't much metadata.

**Comment by weiss** [comment]
> If a client sends a more complex stanza than the strip algorithm can't handle, how do we ensure that the result doesn't leak unexpected content?

Well, my ad-hoc suggestion [above][1] was to "strip all stanza contents except for the stanza's name, its type attribute (if any), and the direct child elements with their xmlns attributes (if any)."  This assumes the element names and `type`/`xmlns` values won't contain sensitive data.  But yes, this does leak some details about the type of data your users are receiving.  (Though to me this seems less sensitive than the user JIDs currently leaked to your app server as per the XEP.)

> it will still leak more metadata to me than I want compared to boiling it down to just `message` and `invite` types.

Yes, if we're reasonably confident that `message` and `invite` will be "enough for anybody"™ then there's no point in my suggestion.  My fear is that other clients might have push requirements we failed to anticipate while predefining notification types.

[1]: https://github.com/ChatSecure/ChatSecure-iOS/issues/637#issuecomment-286808629

**Comment by chrisballinger** [comment]
@weiss 

> (Though to me this seems less sensitive than the user JIDs currently leaked to your app server as per the XEP.)

I also would like that to not be leaked as well, because it's not needed in our case. Perhaps that could be addressed in future revision and made configurable by the client as well.

> My fear is that other clients might have push requirements we failed to anticipate while predefining notification types.

I agree, perhaps we could still boil it down to a few simple types but offer an escape hatch w/ stripped or full stanzas.

**Comment by gerg5c42g542g2c54g52c** [comment]
It would be nice if it was possible to fix push by solely client-side changes as many servers don't use prosody, but use old ejabberd version with pre-alpha mod_push. Also some use prosody 0.10.

**Comment by tmolitor-stud-tu** [comment]
@gerg5c42g542g2c54g52c That won't be possible, sorry...

**Comment by gerg5c42g542g2c54g52c** [comment]
@tmolitor-stud-tu So as I understand this, push on iOS won't work on any ejabberd servers in a reasonably close future as mod_push is the only push plugin for it and it's not actively maintained for the needed changes to be done in it. This could be addes as a simple warning either on the website, appstore description or in the app (check for server version, warn).

**Comment by chrisballinger** [comment]
@gerg5c42g542g2c54g52c What wrong with ejabberd's mod_push? The issue with Prosody 0.10 and ejabberd 17.01 is resolved in v4.0.4, which should be available some time early next week.

**Comment by weiss** [comment]
> It would be nice if it was possible to fix push by solely client-side changes as many servers don't use prosody, but use old ejabberd version with pre-alpha mod_push.

I don't think many servers use the "pre-alpha" module you have in mind (I'm not aware of a single public server that does).

> So as I understand this, push on iOS won't work on any ejabberd servers in a reasonably close future

Depends on your definition of "reasonable".  A proper `mod_push` module will be shipped with one of the next ejabberd releases.

**Comment by chrisballinger** [comment]
Alright, I'm going to close this issue because it's gone way beyond its original scope and now that v4.0.4 is out there are additional resources for end users to self-debug their issues. Please open separate issues for new feature discussion.
